python/source/flight.rst

```diff
@@ -0,0 +1,228 @@
+            self._repo = repo
+
+        def _make_flight_info(self, dataset):
+            dataset_path = self._repo / dataset
+            schema = pa.parquet.read_schema(dataset_path)
+            metadata = pa.parquet.read_metadata(dataset_path)
+            descriptor = pa.flight.FlightDescriptor.for_path(
+                dataset.encode('utf-8')
+            )
+            endpoints = [pa.flight.FlightEndpoint(dataset, [self._location])]
+            return pyarrow.flight.FlightInfo(schema,
+                                            descriptor, 
+                                            endpoints,
+                                            metadata.num_rows, 
+                                            metadata.serialized_size)
+
+        def list_flights(self, context, criteria):
+            for dataset in self._repo.iterdir():
+                yield self._make_flight_info(dataset.name)
+
+        def get_flight_info(self, context, descriptor):
+            return self._make_flight_info(descriptor.path[0].decode('utf-8'))
+
+        def do_put(self, context, descriptor, reader, writer):
+            dataset = descriptor.path[0].decode('utf-8')
+            dataset_path = self._repo / dataset
+            data_table = reader.read_all()
+            pa.parquet.write_table(data_table, dataset_path)
+
+        def do_get(self, context, ticket):
+            dataset = ticket.ticket.decode('utf-8')
+            dataset_path = self._repo / dataset
+            return pa.flight.RecordBatchStream(pa.parquet.read_table(dataset_path))
+
+        def list_actions(self, context):
+            return [
+                ("drop_dataset", "Delete a dataset."),
+            ]
+
+        def do_action(self, context, action):
+            if action.type == "drop_dataset":
+                self.do_drop_dataset(action.body.to_pybytes().decode('utf-8'))
+            else:
+                raise NotImplementedError
+
+        def do_drop_dataset(self, dataset):
+            dataset_path = self._repo / dataset
+            dataset_path.unlink()
+
+The example server exposes :meth:`pyarrow.flight.FlightServerBase.list_flights`
+which is the method in charge of returning the list of data streams available
+for fetching.
+
+Likewise, :meth:`pyarrow.flight.FlightServerBase.get_flight_info` provides
+the information regarding a single specific data stream.
+
+Then we expose :meth:`pyarrow.flight.FlightServerBase.do_get` which is in charge
+of actually fetching the exposed data streams and sending them to the client.
+
+Allowing to list and dowload data streams would be pretty useless if we didn't
+expose a way to create them, this is the responsability of
+:meth:`pyarrow.flight.FlightServerBase.do_put` which is in charge of receiving
+new data from the client and dealing with it (in this case saving it
+into a parquet file)
+
+This are the most common Arrow Flight requests, if we need to add more
+functionalities, we can do so using custom actions.
```

Comment:
Is it worth noting somewhere that if not implemented, all Flight methods have a default implementation? 
Similarly, it may be worth linking to the list of all methods. (Oh, though the docs don't really have a good listing. That should be fixed.)